### PR TITLE
Lazy loading: load static initializers for needed classes. Fixes #846

### LINKED ISSUE #846: CBMC verification failure with --lazy-methods, brace-initialized static array
Using the current tip of test-gen-support (7cbe566), the following program fails verification when invoked with `--lazy-methods` and succeeds otherwise. The full command-line is `cbmc Main.class (--lazy-methods)`.

```java
class Problem {
  private static final Object[] DEFAULT = {};
  private Object data;

  Problem() {
    this.data = DEFAULT;
  }

  void checkInvariant() {
    assert data != null;
  }
}

public class Main {
  public static void main(String[] args) {
    new Problem().checkInvariant();
  }
}
```

@smowton, do you know why this might be?

### PR DESCRIPTION
**Pull request by smowton**
This adds a wrapper for the needed-methods and needed-classes objects used to accumulate dependencies during lazy loading, such that the first time a class is noted needed, its static initializer if any is also queued for elaboration.

The test case is borrowed from @reuk.

### CODE REVIEW

**Comment by tautschnig** [comment]
@smowton could you please take a look at the overly long lines reported?

File: src/java_bytecode/ci_lazy_methods.cpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
+/*******************************************************************\
+
+Module: Context-insensitive lazy methods container
+
+Author: Chris Smowton, chris.smowton@diffblue.com
+
+\*******************************************************************/
+
+#include "ci_lazy_methods.h"
+#include <string>
```
Comment (tautschnig) [inline src/java_bytecode/ci_lazy_methods.cpp]: Nit picking: STL headers first, blank line, local include

File: src/java_bytecode/ci_lazy_methods.cpp [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+void ci_lazy_methodst::add_needed_method(const irep_idt &method_symbol_name)
```
Comment (tautschnig) [inline src/java_bytecode/ci_lazy_methods.cpp]: Function comment headers missing

File: src/java_bytecode/ci_lazy_methods.h [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+#ifndef CPROVER_JAVA_BYTECODE_CI_LAZY_METHODS_H
+#define CPROVER_JAVA_BYTECODE_CI_LAZY_METHODS_H
+
+#include <vector>
+#include <set>
+#include <util/symbol_table.h>
+
+class ci_lazy_methodst
+{
+ public:
```
Comment (tautschnig) [inline src/java_bytecode/ci_lazy_methods.h]: Single-character indent - no indent needed for `public`/`private`/`protected`

File: src/java_bytecode/ci_lazy_methods.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  symbol_table(_symbol_table)
+  {}
+
+  void add_needed_method(const irep_idt &);
+  // Returns true if new
+  bool add_needed_class(const irep_idt &);
+
+ private:
+  std::vector<irep_idt> &needed_methods;
+  std::set<irep_idt> &needed_classes;
```
Comment (tautschnig) [inline src/java_bytecode/ci_lazy_methods.h]: Why is one of those a vector while the other is a set?

**Comment by smowton** [comment]
Amended and added comments to address your question about vector vs. set.

File: src/java_bytecode/ci_lazy_methods.cpp [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+          a static field belonging to it will be accessed. Also notes
+          that its static initializer is therefore reachable.
+
+\*******************************************************************/
+
+bool ci_lazy_methodst::add_needed_class(const irep_idt &class_symbol_name)
+{
+  if(!needed_classes.insert(class_symbol_name).second)
+    return false;
+  const irep_idt &clinit_name=id2string(class_symbol_name)+".<clinit>:()V";
```
Comment (reuk) [inline src/java_bytecode/ci_lazy_methods.cpp]: What's the reason for taking a const ref to a temporary here vs. the following?
```c++
const irep_idt clinit_name(id2string(class_symbol_name)+".<clinit>:()V");
```

File: src/java_bytecode/ci_lazy_methods.h [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  // needed_methods is a vector because it's used as a work-list
+  // which is periodically cleared. It can't be relied upon to
+  // contain all methods that have previously been elaborated.
+  // It should be changed to a set if we develop the need to use
+  // it that way.
+  std::vector<irep_idt> &needed_methods;
+  // needed_classes on the other hand is a true set of every class
+  // found so far, so we can use a membership test to avoid
+  // repeatedly exploring a class hierarchy.
+  std::set<irep_idt> &needed_classes;
```
Comment (tautschnig) [inline src/java_bytecode/ci_lazy_methods.h]: Thanks a lot for the clarification; I'd suggest to use `std::unordered_set<irep_idt, irep_id_hash> &needed_classes` as a minor improvement.

File: src/java_bytecode/ci_lazy_methods.cpp (hunk 5 above)
Comment (smowton) [inline src/java_bytecode/ci_lazy_methods.cpp]: Whoops, fixed

File: src/java_bytecode/ci_lazy_methods.h (hunk 6 above)
Comment (smowton) [inline src/java_bytecode/ci_lazy_methods.h]: I'd want to make change a few others too, so I'll do that in a subsequent PR

**Pull request merged**